File: paper18-hospital-deserts/03_analysis/69_recipient_control_spillover.py
#!/usr/bin/env python
"""
69_recipient_control_spillover.py — flow-construction helper for the
recipient-control spillover bias bound (referee #1 threat).

Builds, from the revealed post-closure admission flows, the set of
never-treated control municipalities that are "recipients" of redirected
psychiatric volume after the 48 PNASH closures. Emits an intermediate
parquet consumed by 69_recipient_control_spillover.R.

Logic (see header of the .R for the economics):
  For each PNASH closing hospital h closing at t*:
    1. flow-exposed munis of h = munis sending admissions to h in
       [t*-W, t*-1] (the patients about to be displaced).
    2. substitutes of h = hospitals (CNES != h) ranked by the increase in
       inflow FROM h's flow-exposed munis, post [t*+1,t*+W] minus pre
       [t*-W,t*-1]. Keep top-K by that delta with positive delta.
    3. for every municipality m, compute its OWN admission SHARE to the
       substitute set, pre vs post. A muni is a recipient (for this closure)
       if delta_share > THRESH.
  A control is recipient-exposed if it qualifies for ANY closure.

Cache-aware: skips if output exists unless --force.
DuckDB out-of-core, 4 threads (parallel job running).
"""
import argparse, os, sys, time, json
import duckdb

ROOT = "/home/darciogm1/projetos/bitter-pills/paper18-hospital-deserts"
EDGES = f"{ROOT}/02_data/intermediate/bipartite_edges.parquet"
PNASH = f"{ROOT}/02_data/processed/pnash_event_level_dataset.parquet"
OUT   = f"{ROOT}/02_data/intermediate/recipient_control_flow.parquet"
LOG   = f"{ROOT}/04_logs/69_recipient_control_spillover_py.log"


def log(msg, fh):
    line = f"[{time.strftime('%H:%M:%S')}] {msg}"
    print(line); fh.write(line + "\n"); fh.flush()


def build(window, top_k, thresh, fh):
    con = duckdb.connect()
    con.sql("PRAGMA threads=4")
    con.sql("PRAGMA memory_limit='12GB'")
    con.sql("PRAGMA temp_directory='/tmp/duckdb_spill'")

    pnash = con.sql(f"SELECT CNES, closure_year FROM read_parquet('{PNASH}')").df()
    log(f"PNASH closures: {len(pnash)} (years {sorted(pnash.closure_year.unique())})", fh)

    # whole-edges in memory view (2.7M rows -> fine)
    con.sql(f"CREATE TEMP VIEW E AS SELECT codmun_6, CNES, year, n_internacoes FROM read_parquet('{EDGES}')")

    all_recip = []          # rows: codmun_6, CNES_closed, delta_share
    subs_audit = []         # rows: CNES_closed, sub_CNES, delta_inflow, rank
    per_closure = []        # diagnostics

    for _, r in pnash.iterrows():
        h, t = r.CNES, int(r.closure_year)
        pre0, pre1 = t - window, t - 1
        post0, post1 = t + 1, t + window

        # 1. flow-exposed munis of h (pre window)
        exp = con.sql(f"""
            SELECT codmun_6, SUM(n_internacoes) v
            FROM E WHERE CNES='{h}' AND year BETWEEN {pre0} AND {pre1}
            GROUP BY codmun_6 HAVING SUM(n_internacoes) > 0
        """).df()
        n_exp = len(exp)
        if n_exp == 0:
            per_closure.append(dict(CNES=h, t=t, n_exposed=0, n_subs=0, n_recip=0))
            continue
        exp_list = "','".join(exp.codmun_6.tolist())

        # 2. substitutes: inflow FROM h's exposed munis, post minus pre, by hospital
        subs = con.sql(f"""
            WITH pre AS (
              SELECT CNES, SUM(n_internacoes) v FROM E
              WHERE codmun_6 IN ('{exp_list}') AND CNES != '{h}'
                AND year BETWEEN {pre0} AND {pre1} GROUP BY CNES),
            post AS (
              SELECT CNES, SUM(n_internacoes) v FROM E
              WHERE codmun_6 IN ('{exp_list}') AND CNES != '{h}'
                AND year BETWEEN {post0} AND {post1} GROUP BY CNES)
            SELECT COALESCE(post.CNES, pre.CNES) AS CNES,
                   COALESCE(post.v,0) - COALESCE(pre.v,0) AS delta
            FROM post FULL OUTER JOIN pre USING (CNES)
            WHERE COALESCE(post.v,0) - COALESCE(pre.v,0) > 0
            ORDER BY delta DESC LIMIT {top_k}
        """).df()
        n_subs = len(subs)
        if n_subs == 0:
            per_closure.append(dict(CNES=h, t=t, n_exposed=n_exp, n_subs=0, n_recip=0))
            continue
        sub_list = "','".join(subs.CNES.astype(str).tolist())
        for rk, (_, sr) in enumerate(subs.iterrows(), 1):
            subs_audit.append(dict(CNES_closed=h, sub_CNES=sr.CNES, delta_inflow=float(sr.delta), rank=rk))

        # 3. for EVERY muni, share to substitute set, pre vs post
        recip = con.sql(f"""
            WITH pre AS (
              SELECT codmun_6,
                     SUM(CASE WHEN CNES IN ('{sub_list}') THEN n_internacoes ELSE 0 END) AS s,
                     SUM(n_internacoes) AS tot
              FROM E WHERE year BETWEEN {pre0} AND {pre1} GROUP BY codmun_6),
            post AS (
              SELECT codmun_6,
                     SUM(CASE WHEN CNES IN ('{sub_list}') THEN n_internacoes ELSE 0 END) AS s,
                     SUM(n_internacoes) AS tot
              FROM E WHERE year BETWEEN {post0} AND {post1} GROUP BY codmun_6)
            SELECT COALESCE(post.codmun_6, pre.codmun_6) AS codmun_6,
                   (CASE WHEN COALESCE(post.tot,0)>0 THEN post.s/post.tot ELSE 0 END) -
                   (CASE WHEN COALESCE(pre.tot,0)>0  THEN pre.s/pre.tot  ELSE 0 END) AS delta_share
            FROM post FULL OUTER JOIN pre USING (codmun_6)
        """).df()
        recip = recip[recip.delta_share > thresh].copy()
        recip["CNES_closed"] = h
        all_recip.append(recip[["codmun_6", "CNES_closed", "delta_share"]])
        per_closure.append(dict(CNES=h, t=t, n_exposed=n_exp, n_subs=n_subs, n_recip=len(recip)))

    import pandas as pd
    recip_df = pd.concat(all_recip, ignore_index=True) if all_recip else \
        pd.DataFrame(columns=["codmun_6", "CNES_closed", "delta_share"])
    subs_df = pd.DataFrame(subs_audit)
    pc_df = pd.DataFrame(per_closure)

    # collapse to muni-level recipient flag (qualifies for ANY closure)
    muni = recip_df.groupby("codmun_6").agg(
        n_closures_recip=("CNES_closed", "nunique"),
        max_delta_share=("delta_share", "max")).reset_index()
    muni["recipient_exposed"] = True

    con.register("muni_df", muni)
    con.sql(f"COPY (SELECT * FROM muni_df) TO '{OUT}' (FORMAT PARQUET, COMPRESSION 'snappy')")
    subs_df.to_parquet(OUT.replace(".parquet", "_subs_audit.parquet"))
    pc_df.to_parquet(OUT.replace(".parquet", "_per_closure.parquet"))

    log(f"closures with >=1 substitute: {(pc_df.n_subs>0).sum()} / {len(pc_df)}", fh)
    log(f"unique recipient munis (any closure): {len(muni)}", fh)
    log(f"wrote {OUT}", fh)
    return len(muni)


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--window", type=int, default=2)
    ap.add_argument("--top_k", type=int, default=3)
    ap.add_argument("--thresh", type=float, default=0.0)
    ap.add_argument("--force", action="store_true")
    a = ap.parse_args()
    os.makedirs(os.path.dirname(LOG), exist_ok=True)
    os.makedirs("/tmp/duckdb_spill", exist_ok=True)
    with open(LOG, "a") as fh:
        log(f"=== run window={a.window} top_k={a.top_k} thresh={a.thresh} force={a.force} ===", fh)
        if os.path.exists(OUT) and not a.force:
            log(f"cache hit {OUT}; skip (use --force)", fh)
            return
        t0 = time.time()
        n = build(a.window, a.top_k, a.thresh, fh)
        log(f"done in {time.time()-t0:.1f}s, {n} recipient munis", fh)


if __name__ == "__main__":
    main()

File: paper18-hospital-deserts/03_analysis/69_recipient_control_spillover.R
#!/usr/bin/env Rscript
# =====================================================================
# 69_recipient_control_spillover.R
# Spillover bias-bound for the staggered DiD mortality null (referee #1).
#
# THREAT: when a psychiatric hospital h closes, displaced patients are
# redirected to a substitute hospital h'. If h' also serves never-treated
# CONTROL municipalities, those controls get partially treated in the SAME
# direction (congestion, disrupted continuity) -> treated-vs-control
# contrast is attenuated TOWARD the null. Donut/health-region dropping is
# necessary but not sufficient; this sharpens it with REVEALED post-closure
# redirection (admission flows), not geography.
#
# TEST: re-estimate the canonical pop-weighted Sun-Abraham ATT for
# suicide_per100k and selfharm_per100k after DROPPING never-treated control
# munis that are revealed RECIPIENTS of redirected volume. If the estimate
# is stable, spillover is not masking a positive effect. If it moves UP,
# that is a finding.
#
# Recipient flags come from 69_recipient_control_spillover.py (DuckDB flow
# construction). This script applies the share-rise threshold, drops
# recipient controls, re-fits, and runs a threshold sensitivity.
#
# Cache-aware (--force). Telemetry per monorepo convention. 4 threads.
# =====================================================================

suppressMessages({
  library(arrow); library(data.table); library(fixest)
})
setDTthreads(4); setFixest_nthreads(4)

ROOT  <- "/home/darciogm1/projetos/bitter-pills/paper18-hospital-deserts"
PANEL <- file.path(ROOT, "02_data/intermediate/staggered_panel_pnash48_ext.parquet")
FLOW  <- file.path(ROOT, "02_data/intermediate/recipient_control_flow.parquet")
OUT   <- file.path(ROOT, "02_data/processed/recipient_control_spillover.csv")
LOG   <- file.path(ROOT, "04_logs/69_recipient_control_spillover.log")

args  <- commandArgs(trailingOnly = TRUE)
FORCE <- "--force" %in% args

PRIMARY_THRESH <- 0.01            # >1pp share rise = revealed recipient (primary)
SENS_THRESH    <- c(0.0, 0.02, 0.05)  # sensitivity to the share-rise cutoff

# --- telemetry --------------------------------------------------------
logf <- function(...) {
  msg <- sprintf("[%s] %s", format(Sys.time(), "%H:%M:%S"), sprintf(...))
  cat(msg, "\n"); cat(msg, "\n", file = LOG, append = TRUE)
}
mem_rss <- function() {
  tryCatch(round(as.numeric(system(sprintf("ps -o rss= -p %d", Sys.getpid()),
           intern = TRUE)) / 1024, 1), error = function(e) NA)
}

if (file.exists(OUT) && !FORCE) {
  logf("cache hit %s; skip (use --force)", OUT); quit(save = "no")
}
cat("", file = LOG, append = TRUE)
logf("=== START host=%s nproc=%s RAM=%s ===",
     Sys.info()[["nodename"]], system("nproc", intern = TRUE),
     gsub("\\n.*", "", system("free -h | awk 'NR==2{print $2}'", intern = TRUE)))
logf("primary thresh=%.2f, sensitivity=%s", PRIMARY_THRESH,
     paste(SENS_THRESH, collapse = ","))

# --- data -------------------------------------------------------------
d <- as.data.table(read_parquet(PANEL))
d[, gn := ifelse(is.na(g_emb) | g_emb == 0, 10000L, g_emb)]
flow <- as.data.table(read_parquet(FLOW))   # codmun_6, n_closures_recip, max_delta_share
logf("panel rows=%d munis=%d | flow recipient candidates=%d | RSS=%sMB",
     nrow(d), uniqueN(d$codmun_6), nrow(flow), mem_rss())

ctrl_munis <- unique(d[gn == 10000L, codmun_6])
treat_munis <- unique(d[gn != 10000L, codmun_6])
logf("treated munis=%d | never-treated controls=%d", length(treat_munis), length(ctrl_munis))

# --- estimator --------------------------------------------------------
fit_att <- function(dat, y) {
  f <- as.formula(sprintf("%s ~ sunab(gn, year) | muni_id + year", y))
  m <- feols(f, dat, cluster = "muni_id", weights = ~pop)
  s <- summary(m, agg = "att")$coeftable
  est <- s[1, 1]; se <- s[1, 2]
  list(att = est, se = se, lo = est - 1.96 * se, hi = est + 1.96 * se,
       p = s[1, 4], nobs = m$nobs)
}

# drop recipient controls at a given threshold; keep ALL treated + clean controls
drop_recipients <- function(dat, thresh) {
  recip <- flow[max_delta_share > thresh, codmun_6]
  recip_ctrl <- intersect(recip, ctrl_munis)   # only controls can be dropped
  list(data = dat[!(gn == 10000L & codmun_6 %in% recip_ctrl)],
       n_recip_ctrl = length(recip_ctrl),
       n_clean_ctrl = length(ctrl_munis) - length(recip_ctrl))
}

# --- baseline (all controls) -----------------------------------------
res <- list()
for (y in c("suicide_per100k", "selfharm_per100k")) {
  b <- fit_att(d, y)
  res[[length(res) + 1]] <- data.table(
    outcome = y, spec = "baseline_all_controls", thresh = NA_real_,
    n_recip_ctrl_dropped = 0L, n_clean_ctrl = length(ctrl_munis),
    att = b$att, se = b$se, ci_lo = b$lo, ci_hi = b$hi, pval = b$p, nobs = b$nobs)
  logf("BASELINE %-16s ATT=%.2f SE=%.2f CI[%.2f,%.2f] N=%d",
       y, b$att, b$se, b$lo, b$hi, b$nobs)
}

# --- primary: drop revealed-recipient controls -----------------------
dp <- drop_recipients(d, PRIMARY_THRESH)
logf("PRIMARY thresh=%.2f: dropped %d recipient controls; %d clean controls remain",
     PRIMARY_THRESH, dp$n_recip_ctrl, dp$n_clean_ctrl)
for (y in c("suicide_per100k", "selfharm_per100k")) {
  b <- fit_att(dp$data, y)
  res[[length(res) + 1]] <- data.table(
    outcome = y, spec = "drop_recipient_controls", thresh = PRIMARY_THRESH,
    n_recip_ctrl_dropped = dp$n_recip_ctrl, n_clean_ctrl = dp$n_clean_ctrl,
    att = b$att, se = b$se, ci_lo = b$lo, ci_hi = b$hi, pval = b$p, nobs = b$nobs)
  logf("DROP-RECIP %-16s ATT=%.2f SE=%.2f CI[%.2f,%.2f] N=%d",
       y, b$att, b$se, b$lo, b$hi, b$nobs)
}

# --- sensitivity to the share-rise threshold -------------------------
for (th in SENS_THRESH) {
  ds <- drop_recipients(d, th)
  for (y in c("suicide_per100k", "selfharm_per100k")) {
    b <- fit_att(ds$data, y)
    res[[length(res) + 1]] <- data.table(
      outcome = y, spec = "sensitivity_thresh", thresh = th,
      n_recip_ctrl_dropped = ds$n_recip_ctrl, n_clean_ctrl = ds$n_clean_ctrl,
      att = b$att, se = b$se, ci_lo = b$lo, ci_hi = b$hi, pval = b$p, nobs = b$nobs)
  }
  logf("SENS thresh=%.2f dropped=%d clean=%d | suicide ATT=%.2f selfharm ATT=%.2f",
       th, ds$n_recip_ctrl, ds$n_clean_ctrl,
       res[[length(res) - 1]]$att, res[[length(res)]]$att)
}

# --- write ------------------------------------------------------------
out <- rbindlist(res)
num <- c("att", "se", "ci_lo", "ci_hi", "pval", "thresh")
out[, (num) := lapply(.SD, function(x) round(x, 2)), .SDcols = num]
fwrite(out, OUT)
logf("wrote %s (%d rows) | RSS=%sMB | DONE", OUT, nrow(out), mem_rss())
print(out)
